### ISSUE TITLE
Layerwise Hessian matrices

### ISSUE THREAD

**Issue by mmuratarat** [gknilsen/pyhessian#1]
First of all, thank you for this very useful function. I have a question. How can we derive layer-wise Hessian matrices from full Hessian matrix?

**Comment by gknilsen** [comment]
Hello, and thank you for your response. I am working on this and more, it will be released shortly.

**Comment by mmuratarat** [comment]
Thank you for your response, Dr Nilsen. Since we have full hessian matrix, wouldn't splitting the Hessian, according to parameters since we flatten them, do the job?

**Comment by gknilsen** [comment]
Yes, this is what I am doing. However, in large nets, even Hessian for a single layer can be intractable in terms of size. Thoughts around that?

**Comment by mmuratarat** [comment]
Yes exactly, I have not thought about it pretty much, to be honest.

In large networks, The Hessian matrix will be huge and therefore, I am trying to see whether LeCun's approximation to find eigenvalues/eigenvectors without computing Hessian matrix works or not. 

My sloppy solution to split the Hessian is something like this:
Since you already gather weights and flatten them, the columns of Hessian will be ordered anyway (it is also symmetric). 

```
n_features, N2, n_classes= 16, 10, 3
...
params = [W1, b2, W2, b3]
....
#First set of weights belonging to the first layer
H[0:n_features*N2,0:n_features*N2]))
#The biases belonging to the first layer
H[n_features*N2:n_features*N2+N2,n_features*N2:n_features*N2+N2]))
#Second set of weights belonging to the output layer
H[n_features*N2+N2:(n_features*N2+N2)+(N2*n_classes), n_features*N2+N2:(n_features*N2+N2)+(N2*n_classes)]))
#The biases belonging to the output layer
 H[(n_features*N2+N2)+(N2*n_classes):H.shape[0], (n_features*N2+N2)+(N2*n_classes):H.shape[0]]))
```
And I could manage to compute the same with `tf.hessian()` function of Tensorflow.

**Comment by gknilsen** [comment]
Your Hessian will have the following structure;
![image](https://user-images.githubusercontent.com/42994401/58248723-26544c00-7d5d-11e9-9577-fdcda9d817fc.png)

**Comment by mmuratarat** [comment]
Yes, this is exactly what i was thinking! Thank you, Dr Nilsen. Tensorflow can only give the matrices that you named as W1, b2 and W2 and b3 on the image you provided (which are matrices on the diagonal).

**Comment by gknilsen** [comment]
Yes. I should have named the matrices in the image otherwise btw, not to directly confuse with the original weight matrices/bias vectors. 
What is the LeCun eigen approximation reference? Is it possible to cheaply find a few Hessian eigen vectors / values and this way approximate the Hessian/Inverse Hessian?

**Comment by mmuratarat** [comment]
Yes I totally understood what is what, Dr Nilsen.

well it is used to approximate eigenvalues without computing hessian matrix. 
http://yann.lecun.com/exdb/publis/pdf/lecun-simard-pearlmutter-93.pdf
http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf

But I cannot make it work yet.

**Comment by gknilsen** [comment]
Thanks!
